llvm/lib/CodeGen/TargetOptionsImpl.cpp

```diff
@@ -40,6 +40,20 @@ bool TargetOptions::DisableFramePointerElim(const MachineFunction &MF) const {
   llvm_unreachable("unknown frame pointer flag");
 }
 
+/// DisableLeafFramePointerElim - This returns true if leaf frame pointer
+/// elimination optimization should be disabled for the given machine function.
+bool TargetOptions::DisableLeafFramePointerElim(
```

Comment:
DisableFramePointerElim itself should be sufficient; it already checks whether the function is a leaf function.

Reply:
Yes, you're right, but the logic in determineCalleeSaves currently does not align with this scenario, which involves saving frame pointer (FP) registers with -mno-omit-leaf-frame-pointer and -fno-omit-frame-pointer
https://github.com/llvm/llvm-project/blob/9b5a3036adbd9eed0f377cdf06aa622b917a1699/llvm/lib/Target/ARM/ARMFrameLowering.cpp#L2265-L2273
I would appreciate any suggestions for improvements or modifications you might have. Thank you!

Reply:
I mean, DisableFramePointerElim() should return true in all the cases where we want to disable frame-pointer elimination.  So you shouldn't need a separate API specifically for leaf functions.  I think if you just replace the call to DisableLeafFramePointerElim with a call to DisableFramePointerElim, you should get the same behavior.  So the target-independent code shouldn't need changes.

Probably at that point, the logic could be refactored a bit.

Reply:
Okay, thank you for your advice. I'll remove `DisableFramePointerElim()` and refactor CanEliminateFrame like this:
`bool CanEliminateFrame = !(requiresAAPCSFrameRecord(MF) && hasFP(MF)) ||
      !(MF.getFunction().hasFnAttribute("frame-pointer") && MF.getFunction().getFnAttribute("frame-pointer").getValueAsString() == "all");`

Reply:
I don't quite follow this... I'd prefer to continue using the target-independent DisableFramePointerElim() where possible.

Reply:
I did an experiment and changed it like this `bool CanEliminateFrame = !(requiresAAPCSFrameRecord(MF) && hasFP(MF)) && !MF.getTarget().Options.DisableFramePointerElim(MF);`
However, fp is retained in the following scenarios:
https://github.com/llvm/llvm-project/blob/3f37c517fbc40531571f8b9f951a8610b4789cd6/llvm/lib/Target/ARM/ARMFrameLowering.cpp#L180-L185
I don't know if it makes sense. For example, the fast-isel-intrinsic.ll test case saves fp.

Reply:
like this,  
```_t1:
@ %bb.0:
	push	{r7, lr}
	mov	r7, sp
	movw	r0, :lower16:_message1
	movt	r0, :upper16:_message1
	add	r0, r0, #5
	movw	r1, #64
	movw	r2, #10
	and	r1, r1, #255
	bl	_memset
	pop	{r7, pc}
                                        @ -- End function
	.globl	_t2                             @ -- Begin function t2
	.p2align	2
	.code	32                              @ @t2
```
Because this judgment is at the front
https://github.com/llvm/llvm-project/blob/22829f757dc76b23071d9438ae9c6ddc3e966db0/llvm/lib/CodeGen/TargetOptionsImpl.cpp#L24-L31

Reply:
If you want to preserve the existing behavior, maybe ARMFrameLowering::keepFramePointer() should check MF.getFrameInfo().hasCalls()?

Reply:
Actually, hmm.  I guess the way this works on ARM specifically is that we have a state between "non-leaf" and "leaf": basically, we have a frame pointer if we spill any registers.  (Or a few other obscure reasons which are mostly irrelevant outside of M-class targets.)  And currently, the ARM target translates "leaf" to this.

I think the best thing to do would be to get rid of ARMFrameLowering::keepFramePointer(), and add the useFastISel() check to ARMFrameLowering::hasFP() instead.  That should preserve the behavior we want here, I think.

Reply:
thank you, I've made the code modifications. I hope you can help me check it.